Framework/DataHandling/src/LoadMask.cpp

```diff
@@ -39,6 +39,234 @@ using namespace Mantid::Kernel;
 using namespace Mantid::API;
 using namespace std;
 
+namespace {
+// service routines
+//-------------------------------------------------------------------------------------------
+/** Convert ranged vectors to single-valued vector
+* @param singles -- input vector of single numbers to copy to result without
+*                   changes.
+* @param ranges  -- input vector of data ranges -- pairs of min-max values,
+*                   expanded to result as numbers from min to max
+*                   inclusively, with step 1
+* @param tot_singles -- on input contains range of single values already
+*                   copied into the result by previous call to the routine,
+*                   on output, all signles and expanded pairs
+*                   from the input are added to it.
+*/
+template <typename T>
+void convertToVector(const std::vector<T> &singles,
+                     const std::vector<T> &ranges,
+                     std::vector<T> &tot_signles) {
+
+  // find the size of the final vector of masked values
+  size_t n_total(singles.size() + tot_signles.size());
+  for (size_t i = 0; i < ranges.size(); i += 2) {
+    n_total += ranges[i + 1] - ranges[i] + 1;
+  }
+  // reserve space for all masked spectra
+  // for efficient memory operations
+  tot_signles.reserve(n_total);
+  // add singles to the existing singles
+  tot_signles.insert(tot_signles.end(), singles.begin(), singles.end());
+  // expand pairs
+  for (size_t i = 0; i < ranges.size(); i += 2) {
+    for (T obj_id = ranges[i]; obj_id < ranges[i + 1] + 1; ++obj_id) {
+      tot_signles.push_back(obj_id);
+    }
+  }
+}
+
+/*
+* Parse index range text to singles and pairs
+* Example: 3,4,9-10,33
+*
+* @param inputstr -- input string to process in the format as above
+* @param sinvles -- vector of obects, defined as singles
+* @param pairs   -- vector of obects, defined as pairs, in the form min,max
+*                   value
+*/
+template <typename T>
+void parseRangeText(const std::string &inputstr, std::vector<T> &singles,
+                    std::vector<T> &pairs) {
```

Comment:
I guess you are right here. I was not going to force this anyway. Maybe it would be preferable to find all these local little parsers scattered around the Mantid codebase and unify them to a general one, but that is indeed an issue of another time.